dace/autodiff/utils.py

```diff
@@ -0,0 +1,977 @@
+
+        # extract the in statement only
+        else_statement = else_statement.replace("else ", "")
+
+        # remove the last closing parenthesis if it exists
+        if else_statement.endswith(")"):
+            else_statement = else_statement[:-1]
+
+        # match the out connector
+        matches = re.search(r"^(.)* =", tasklet_code)
+        if not matches:
+            raise AutoDiffException(f"Could not find output assignment in tasklet code: {tasklet_code}")
+        out_connector = matches.group()
+
+        # extract the in statement only
+        out_connector = out_connector.replace(" =", "")
+
+    # sanity check this should be in the out connectors of the tasklet
+    if out_connector not in tasklet_node.out_connectors:
+        raise AutoDiffException(
+            f"Output connector '{out_connector}' not found in tasklet output connectors: {list(tasklet_node.out_connectors.keys())}"
+        )
+
+    # create the return expressions
+    if_expression = f"{out_connector} = {if_statement}"
+    else_expression = f"{out_connector} = {else_statement}" if else_statement else None
+
+    return if_expression, else_expression, conditional
+
+
+def check_edges_type_in_state(subgraph: dstate.StateSubgraphView) -> None:
+    """
+        Check if all the edges in this state are of type float, int, or boolean.
+        """
+    for edge, parent_subgraph in subgraph.all_edges_recursive():
+        if isinstance(parent_subgraph, SDFGState):
+            parent_sdfg = parent_subgraph.parent
+        elif isinstance(parent_subgraph, dstate.StateSubgraphView):
+            parent_sdfg = parent_subgraph.graph.parent
+        elif isinstance(parent_subgraph, SDFG) or isinstance(parent_subgraph, LoopRegion):
+            # if there are any fancy things on the interstate edges we should probably throw an error
+            continue
+        else:
+            raise AutoDiffException("Unexpected subgraph structure")
+
+        if edge.data.data:
+            edge_type = parent_sdfg.arrays[edge.data.data].dtype
+            if edge_type in [dace.string]:
+                raise AutoDiffException(
+                    f"Expected Subgraph to differentiate to only contain float, int, and bool edges, but data {edge.data}"
+                    f" on edge {edge} has type {edge_type}")
+
+
+def state_within_loop(forward_state: SDFGState) -> Tuple[bool, LoopRegion]:
+    """
+    Check if this state will be executed several times within a loop.
+    We check if any of the parents of this state is a loop region.
+    """
+    parent = forward_state.parent_graph
+    while parent is not None:
+        if isinstance(parent, LoopRegion):
+            return True, parent
+        parent = parent.parent_graph
+    return False, None
+
+
+class SympyCleaner(ast.NodeTransformer):
+
+    def visit_Name(self, node):
+        if node.id == "pi":
+            return ast.copy_location(ast.parse("(3.141592653589)").body[0], node)
```

Comment:
Don't we have pi in the runtime library (dace:math:pi)? I think you can use that instead of the constant here.